xdsl/passes.py

```diff
@@ -103,6 +104,35 @@ def from_pass_spec(cls: type[ModulePassT], spec: PipelinePassSpec) -> ModulePass
         return cls(**arg_dict)
 
 
+def _empty_callback(
+    previous_pass: ModulePass, module: builtin.ModuleOp, next_pass: ModulePass
+) -> None:
+    return
+
+
+@dataclass
+class PassPipelinePass(ModulePass):
+    passes: list[ModulePass]
+    callback: Callable[[ModulePass, builtin.ModuleOp, ModulePass], None] = field(
+        default=_empty_callback
+    )
+    """
+    Function called in between every pass, taking the pass that just ran, the module, and
+    the next pass.
+    """
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        if not self.passes:
+            # Early exit to avoid fetching a non-existing last pass.
+            return
+
+        for prev, next in zip(self.passes[:-1], self.passes[1:]):
+            prev.apply(ctx, op)
+            self.callback(prev, op, next)
+
+        self.passes[-1].apply(ctx, op)
```

Comment:
A pipeline by definition is a serial processing regime.
Why do we need to know the next pass?
Can't the callback points only be modelled around a single pass' behaviour?
For example,
- before execution
- after execution
- when the IR changes? (not sure how viable this is)

Reply:
My main idea was that the API guarantees that it's in between passes and not before the first one or after the last one. If you have both then it's unambiguous. The other thing is that it may be useful to execute something before a certain pass, maybe as a debug step, independent of the passes before. There may be other use-cases that I'm not thinking of.

Reply:
Thanks for the clarification; if the handling is at the "pipeline level", this makes sense.